Azure Quantum SDK Readme (#18296)
Azure/azure-sdk-for-net: sdk/quantum/Azure.Quantum.Jobs/README.md

# Azure Quantum Jobs client library for .NET

> Azure Quantum is a Microsoft Azure service that you can use to run quantum computing programs or solve optimization problems in my cloud. Using the Azure Quantum tools and SDKs, you can create quantum programs and run them against different quantum simulators and machines.

Reviewer: ```suggestion
Azure Quantum is a Microsoft Azure service that you can use to run quantum computing programs or solve optimization problems in the cloud.  Using the Azure Quantum tools and SDKs, you can create quantum programs and run them against different quantum simulators and machines.  You can use the Azure.Quantum.Jobs client library to
```

- Create, enumerate, and cancel quantum jobs
- Enumerate provider status and quotas

>   [Source code][source] | [Package (NuGet)][package] | [API reference documentation](https://docs.microsoft.com/python/api/overview/azure/batch?view=azure-python) | [Product documentation](https://docs.microsoft.com/azure/batch/)

Reviewer: Either point the links at your service or remove them.

> ## Getting started

Reviewer: The line below this, that github won't let me comment on, is an instruction for you and not the customer.  It'd be good to remove it.

Then you'll want to copy "Install the Package" from another service to show the instructions for NuGet.  Be sure to include the `--prerelease` with version number or else it won't find anything.

You should update the Prereqs to those required for your service.  I don't think you require Cosmos or Python, for example.

> You must have an [Azure subscription](https://azure.microsoft.com/free/), [Cosmos DB account](https://docs.microsoft.com/azure/cosmos-db/account-overview) (SQL API), and [Python 3.6+](https://www.python.org/downloads/) to use this package.

> ### Authentication

Reviewer: Changing the names of some of these sections will break DocWarden that runs as part of the CI.  I believe "Authenticate the client" is required.

> To authenticate with the service, the workspace will use [DefaultAzureCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet) internally. This will try different authentication mechanisms based on the environment (e.g. Environment Variables, ManagedIdentity, CachedTokens) and finally it will fallback to [InteractiveBrowserCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.interactivebrowsercredential?view=azure-dotnet).

Reviewer: Nit - remove all the "/en-us/" parts of your links.  Docs will add them back when customers visit but make it specific to their default locale.

```suggestion
Use [DefaultAzureCredential](https://docs.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet) to authenticate.  This will try different authentication mechanisms based on the environment (e.g. Environment Variables, Managed Identity, Interactive Prompts, etc.).  See [Azure.Identity](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/README.md) to learn more.
```

> Workspace will also allow the user to override the above behavior by passing their own [TokenCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokencredential?view=azure-dotnet).

Reviewer: Maybe remove this line?  It's not much of a getting started scenario.

> In case the user already has an access token (using MSAL manually or even an external library), we will add an overload that accepts the access token string.

Reviewer: I'd remove this because it doesn't help users get started if you're talking about future plans and we're looking at an OnBehalfOfToken/StaticToken in Azure.Identity that would be preferable to another overload.

> `TokenCredential` is the default Authentication mechanism used by Azure SDKs.

Reviewer: I'd remove this line too.

## Key concepts

> 'QuantumJobClient' is the root class to be used to authenticate and create, enumerate, and cancel jobs.

Reviewer: Use backticks, not quotes.  Ditto for all the code in this section.
```suggestion
`QuantumJobClient` is the root class to be used to authenticate and create, enumerate, and cancel jobs.
```

'JobDetails' contains all the properties of a job.

'ProviderStatus' contains status information for a provider.

'QuantumJobQuota' contains quota properties.

## Examples

* [Get Container SAS URI](#get-container-sas-uri)
* [Upload Input Data](#upload-input-data)
* [Create The Job](#create-the-job)
* [Get Job](#get-job)
* [Get Jobs](#get-jobs)

### Create the client

Create an instance of the QuantumJobClient by passing in these parameters:

> - [Subscription][subscriptions] - looks like XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX and can be found in your list of subscriptions on azure

Reviewer: ```suggestion
- [Subscription Id][subscriptions] - a uuid like XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX found in your list of subscriptions on Azure
```

- [Resource Group][resource-groups] - a container that holds related resources for an Azure solution 
- [Workspace][workspaces] - a collection of assets associated with running quantum or optimization applications

> - [Location][location] - choose the best data center by geographical region 

Reviewer: Is this scoped during preview?  Might be helpful to list the currently valid locations.

> - [StorageContainerName][blob-storage] - your blob storage 

Reviewer: ```suggestion
- [StorageContainerName][blob-storage] - the name of an Azure Storage blob container
```

> - [Credential][credentials] - used to authenticate 

Reviewer: Are their any service limitations on what's allowed?  Like is it only MSI or can any Service Principal be configured?

        credential);
```

### Get Container SAS URI

> Create a storage container where to put your data.

Reviewer: This isn't actually creating a container is it?  I assume it just generates a SAS URI based on the name.

```C# Snippet:Azure_Quantum_Jobs_GetContainerSasUri
// Get container Uri with SAS key
var containerUri = (quantumJobClient.GetStorageSasUri(
    new BlobDetails(storageContainerName))).Value.SasUri;
```

### Upload Input Data

> Using the SAS URI, upload the json input data to the blob client.

Reviewer: ```suggestion
Upload the local `problem.json` file with input data to Azure Storage using the container's SAS Uri.
```

> This contains the parameters to be used with [Quantum Inspired Optimizations](https://github.com/MicrosoftDocs/quantum-docs-private/wiki/Optimization-quickstart-for-Azure-Quantum)

Reviewer: quantum-docs-private in the URI makes me think people won't have access to these?

blobClient.Upload(problemFilename, overwrite: true);
```

### Create The Job

> Now that you have the blob input data initialized, fill in the job fields and pass it to the CreateJob() function.

Reviewer: ```suggestion
Now that you've uploaded your problem definition to Azure Storage, you can use `CreateJob` to define an Azure Quantum job.
```

JobDetails createdJob = (quantumJobClient.CreateJob(jobId, createJobDetails)).Value;
```

### Get Job

> To retrieve a specific job by its job id, you can use this function.

Reviewer: ```suggestion
`GetJob` retrieves a specific job by its id.
```

```C# Snippet:Azure_Quantum_Jobs_GetJob
// Get the job that we've just created based on its jobId
JobDetails myJob = (quantumJobClient.GetJob(jobId)).Value;
```

### Get Jobs

> To enumerate all the jobs in the workspace, use the GetJobs() function.

Reviewer: ```suggestion
To enumerate all the jobs in the workspace, use the `GetJobs` method.
```

```C# Snippet:Azure_Quantum_Jobs_GetJobs
foreach (JobDetails job in quantumJobClient.GetJobs())
{
    Console.WriteLine($"{job.Name}");
}
```

> ## Next steps

Reviewer: Delete everything that isn't an actual next step.  Maybe say something like "Visit our documentation at http://xyz.com to learn more about Azure Quantum."?

## Contributing

See the [CONTRIBUTING.md][contributing] for details on building,
testing, and contributing to this library.

This project welcomes contributions and suggestions.  Most contributions require
you to agree to a Contributor License Agreement (CLA) declaring that you have
the right to, and actually do, grant us the rights to use your contribution. For
details, visit [cla.microsoft.com][cla].

This project has adopted the [Microsoft Open Source Code of Conduct][coc].
For more information see the [Code of Conduct FAQ][coc_faq]
or contact [opencode@microsoft.com][coc_contact] with any
additional questions or comments.

<!-- LINKS -->
[source]: https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/quantum/Azure.Quantum.Jobs/src
[package]: https://www.nuget.org/packages/Azure.Quantum.Jobs/

> [resource-groups]: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/manage-resource-groups-portal

Reviewer: Remove all the `/en-us/`s down here too.
